diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 0362153..1df2908 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -257,10 +257,10 @@ target_link_libraries(ada PUBLIC stb glob skymodel )
 # source_group(   TREE ${INCLUDE_FOLDER} 
 #                 PREFIX "Header Files" FILES ${ROOT_HEADER})
 
-# install(TARGETS ${PROJECT_NAME} DESTINATION lib)
-# install(FILES ${ROOT_HEADER} DESTINATION include/ada)
-# install(FILES ${GL_HEADER} DESTINATION include/ada/gl)
-# install(FILES ${SCENE_HEADER} DESTINATION include/ada/gl)
-# install(FILES ${TOOLS_HEADER} DESTINATION include/ada/tools)
-# install(FILES ${SHADERS_HEADER} DESTINATION include/ada/shaders)
+install(TARGETS ${PROJECT_NAME} DESTINATION lib)
+install(FILES ${ROOT_HEADER} DESTINATION include/ada)
+install(FILES ${GL_HEADER} DESTINATION include/ada/gl)
+install(FILES ${SCENE_HEADER} DESTINATION include/ada/scene)
+install(FILES ${TOOLS_HEADER} DESTINATION include/ada/tools)
+install(FILES ${SHADERS_HEADER} DESTINATION include/ada/shaders)
 
